### ISSUE TITLE
Center location important

### ISSUE THREAD

**Issue by Hendrik-code** [MIC-DKFZ/nnUNet#1609]
Hey,

thanks for the great work in this repo!
So I have a dataset, in which the GT labels are dependent on the center of the image (think of an CT scan cutout, and I want to predict the structure in the center of the cutout)

So obviously, I disabled mirroring, and set the patch size to the (small) image size so there is no random cutout, but always one where the center is still the center (I hope this makes sense). Even though the segmentation task is fairly easy, the results are quite bad.

Is there anything else to consider? Maybe something in the Dataloader3D to ensure that both during training and inference, the center of my input images are always also the center when passed into the model? (and no augmentation ruining this somehow). There are similar structure around the center structure in the input image.

Thanks in advance!

**Comment by sten2lu** [comment]
Hi @Hendrik-code,

your dataset sounds like an interesting case.

I guess you mean that the images are not supposed to be cropped (this is when only a small part of the image is given to the model).

Therefore the relation of the patch size nnU-Net uses for training and the image size is important.
As nnU-Net resamples the data by default to fits its purpose the hackiest way would be to train the lowresolution version of nnU-Net and check how whether it performs better than your current version.

For this I refer to the [documentation](https://github.com/MIC-DKFZ/nnUNet/blob/master/documentation/how_to_use_nnunet.md).

Alternatively/additionaly you could also check the plans files `nnUNetPlans.json` located in the `nnUNet_preprocessed/DatasetXXX` folder.
There you will find a dictionary in which you can compare the value `patch size` and `median_image_size_in_voxels` for each trainer (3d_lowres, 3d_fullres...).
If the patch_size is (roughly) in the same size as the median_image_size_invoxels the model should almost always be able to see the center during training making this trainer save to use.

patch_size is here the input size of the model and median_image_size_invoxels is the resolution to which the images are on average downsampled.


It can also be that a bit of performance is lost due to nnU-Net cropping patches of the image during training where the center is not always present. However, I would advise you to try out the other steps beforehand and make sure that the patch size the model uses is in the same size as the image size.

Best regards
Carsten

**Comment by Hendrik-code** [comment]
Thanks for the answer!

Oh forgot to mention, I also disabled the resampling for this training procedure (and it shouldn't even matter because all my input images are already resampled to the same size). And my patch size in the plans file is equal to the image size of the input images (which are all of the same exact size, resolution, ...) So from that standpoint, the image should always have the exact same size than the patch size used, no?

I am not sure if I understood you correctly, but the center is always present during training I believe. But I believe that there is some mechanism so that the center of the input image is not the center anymore when handed to the model. And for my prediction, the center is the only context of the image (not the actual input data in the center).

Lets say that I put a circle in the center of the images. Regardless of the input images, as long as they are the same size and the patch size is equal to the image size, I would imagine that the model just learns to draw a circle in the middle of the input image, regardless of the input. This is basically not the case for me right now. So I assume during training, that the center somehow moved during its travel from being the input image to the actual model.

Could you explain why 3d_lowres is an option here? Because I thought if 3d_fullres can already cover the entire image, 3d_lowres is kinda out. And at least from inference, I know there is always only one patch processed, so on inference, I 
dont see any problems. I just think that something in the training went wrong, because the results are so bad to the point where the model thinks the center is way above or below the actual center.

Example images of input and expected output (output of course only the green section)
Blue line is the center of the image. as you can see, there are similar structures around (and my actual data is a larger cutout, so neighboring structures are visible) but the model fails to segment the center structure way too often. Am I missing something?

Thanks!

![1](https://github.com/MIC-DKFZ/nnUNet/assets/19862882/0be2085c-95f5-45e4-9193-a4d8403c31e4)
![2](https://github.com/MIC-DKFZ/nnUNet/assets/19862882/5e14d276-0687-4b65-93f6-613dffff2a5e)

**Comment by sten2lu** [comment]
Hi @Hendrik-code,

As the patch size is equal to the image size for the 3d_fullres configuration no further processing would be required.
Therefore the center should always be present during the training.
Also  3d_lowres therefore, should not solve this problem, as there is no additional global information the model gains over 3d_fullres.

A U-Net based on convolutions is designed to be equivariant in it outputs (similar patterns should be detected in every area). 
Therefore the model not being able to detect the center areas might be just because it has seen similar structures that were not in center and therefore unlabeled.

Approaches to solve your problem might be:
1. Adding a new channel to your data which has values that correspond to the distance of the center.
2. Label every similar structure in your dataset to the one you want to be annotated by nnU-Net and select the center structure in postprocessing.


Best regards,

Carsten

**Comment by Hendrik-code** [comment]
Hey @sten2lu ,

Thanks for the answer!
Well, the center is always visible, but what I am saying is that the center when fed during training and/or inference is not the center anymore (like the positioning).

So I just ran a simple convolutional unet network (not using any nnunet code) on the exact same data, and get like a >0.99 Dice after ~100 epochs (still coded as 250 iterations each), equal batch size, patch size, etc... without any data augmentation at all. So the task is actually very easy, as expected.

So *something* must be going on in the nnunet code that explains why there I only get like ~0.7 dice score. And the easiest explanation would be (considering its the exact same training data, ...) is that the center position of the training data gets somehow shifted/translated during training. Don't you agree?

Your other two options are neat, but wont work very well, as the structure might touch each other (so 2. not working). But your first idea is actually brilliant. Still, I think fixing the problem of why the center position is not consistent would be the better solution. (of course, now that I have a model that works perfectly well means I don't really need the nnUnet version anymore, but I still want to understand the cause here).

Best regards,
Hendrik

**Comment by sten2lu** [comment]
Hi @Hendrik-code,

thank you for coming back!
I checked the code and corresponded with Fabian regarding the center location (or also where cropping is applied).

During inference (prediction), the center location stays the same if the size of the image is identical/smaller than the patch size.
However, during training, the center location of the patches that are used to train nnU-Net is not identical to the image center. This is due to the oversampling technique used by nnU-Net. 
Currently, the function which selects the patches is the `get_bbox` method that gets used to obtain these patches (by cropping the image) for the training. Therefore to achieve the desired behavior a new dataloader for training which overrides this method would need to be created. 

I hope this answers the remaining open points.

Best regards,
Carsten

**Comment by Hendrik-code** [comment]
Hey @sten2lu ,

thank you! That makes sense. Even though oversampling (in my opinion) doesn't make sense if the patch size is equal to the image size, but yeah, if you always use the center of the bbox then that could not be the image center, true. 

Best regards,
Hendrik
